plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java

```diff
@@ -265,6 +268,52 @@ long getLargeBlobThresholdInBytes() {
             {Storage.BlobWriteOption.doesNotExist(), Storage.BlobWriteOption.md5Match()};
     private static final Storage.BlobWriteOption[] OVERWRITE_CHECK_MD5 = {Storage.BlobWriteOption.md5Match()};
 
+    void writeBlob(String blobName, boolean failIfAlreadyExists, CheckedConsumer<OutputStream, IOException> writer) throws IOException {
+        final BlobInfo blobInfo = BlobInfo.newBuilder(bucketName, blobName).build();
+        final Storage.BlobWriteOption[] writeOptions = failIfAlreadyExists ? NO_OVERWRITE_NO_MD5 : OVERWRITE_NO_MD5;
+
+        StorageException storageException = null;
+
+        for (int retry = 0; retry < 3; ++retry) {
+            try {
+                final WriteChannel writeChannel = SocketAccess.doPrivilegedIOException(() -> client().writer(blobInfo, writeOptions));
+                try (OutputStream out = new FilterOutputStream(Channels.newOutputStream(new WritableBlobChannel(writeChannel))) {
+                    @Override
+                    public void write(byte[] b, int off, int len) throws IOException {
+                        int written = 0;
+                        while (written < len) {
+                            // at most write the default chunk size in one go to prevent allocating huge buffers in the SDK
+                            // see com.google.cloud.BaseWriteChannel#DEFAULT_CHUNK_SIZE
+                            final int toWrite = Math.min(len - written, 60 * 256 * 1024);
+                            out.write(b, off + written, toWrite);
+                            written += toWrite;
+                        }
+                    }
+                }) {
+                    writer.accept(out);
+                    SocketAccess.doPrivilegedVoidIOException(writeChannel::close);
```

Comment:
This is all we needed for a solution to partial GCP writes in #72051 really. There are no resources attached to the write channel in the SDK, simply not closing it prevents the resumable write from finalizing. I had to go with this approach here instead of the md5 validation because we obviously don't know the hash beforehand when serializing on the fly.